File: forge/src/main/java/juuxel/adorn/platform/forge/compat/Compat.java
package juuxel.adorn.platform.forge.compat;

import juuxel.adorn.block.variant.BlockVariantSets;
import juuxel.adorn.config.ConfigManager;
import juuxel.adorn.platform.forge.registrar.ForgeRegistrar;
import net.neoforged.bus.api.IEventBus;
import net.neoforged.fml.ModList;

public final class Compat {
    public static void init(IEventBus modBus) {
        ifModLoaded("biomemakeover", () -> BlockVariantSets.add(new BiomeMakeoverCompat()));
        ifModLoaded("biomesoplenty", () -> BlockVariantSets.add(new BiomesOPlentyCompat()));
        ifModLoaded("biomeswevegone", () -> BlockVariantSets.add(new BwgCompat()));
        ifModLoaded("ecologics", () -> BlockVariantSets.add(new EcologicsCompat()));
        ifModLoaded("architects_palette", () -> BlockVariantSets.add(new ArchitectsPaletteCompat()));

        ((ForgeRegistrar<?>) BlockVariantSets.BLOCKS).hook(modBus);
        ((ForgeRegistrar<?>) BlockVariantSets.ITEMS).hook(modBus);
    }

    public static boolean isCompatEnabled(String mod) {
        var compatMap = ConfigManager.config().compat;

        if (!compatMap.containsKey(mod)) {
            compatMap.put(mod, true);
            ConfigManager.get().save();
            return true;
        }

        return compatMap.get(mod);
    }

    private static void ifModLoaded(String mod, Runnable fn) {
        if (isCompatEnabled(mod) && ModList.get().isLoaded(mod)) {
            fn.run();
        }
    }
}

File: forge/src/main/kotlin/juuxel/adorn/platform/forge/compat/Compat.kt
package juuxel.adorn.platform.forge.compat

import juuxel.adorn.compat.CompatBlocks
import juuxel.adorn.config.ConfigManager
import juuxel.adorn.platform.forge.registrar.ForgeRegistrar
import net.minecraftforge.eventbus.api.IEventBus
import net.minecraftforge.fml.ModList

object Compat {
    fun init(modBus: IEventBus) {
        val mods = mapOf(
            "biomemakeover" to { CompatBlocks.addVariants(BiomeMakeoverCompat) },
            "biomesoplenty" to { CompatBlocks.addVariants(BiomesOPlentyCompat) },
            "byg" to { CompatBlocks.addVariants(BygCompat) },
            "ecologics" to { CompatBlocks.addVariants(EcologicsCompat) },
            "architects_palette" to { CompatBlocks.addVariants(ArchitectsPaletteCompat) },
        )

        for ((mod, fn) in mods) {
            ifModLoaded(mod, fn)
        }

        CompatBlocks.register()
        (CompatBlocks.blocks as ForgeRegistrar<*>).hook(modBus)
        (CompatBlocks.items as ForgeRegistrar<*>).hook(modBus)
    }

    fun isCompatEnabled(mod: String): Boolean {
        val compatMap = ConfigManager.config().compat

        if (mod !in compatMap) {
            compatMap[mod] = true
            ConfigManager.INSTANCE.save()
            return true
        }

        return compatMap[mod]!!
    }

    private inline fun ifModLoaded(mod: String, fn: () -> Unit) {
        if (isCompatEnabled(mod) && ModList.get().isLoaded(mod)) {
            fn()
        }
    }
}
